File: homework3/src/main/java/com/chocomiruku/homework3/task2/Action.java
package com.chocomiruku.homework3.task2;

public abstract class Action {
    private Action() {
    }

    public static final class Registration extends Action {
    }

    public static final class Login extends Action {
        private final User user;

        public Login(User user) {
            this.user = user;
        }

        public User getUser() {
            return user;
        }
    }

    public static final class Logout extends Action {
    }
}

File: app/src/main/java/com/chocomiruku/homework1/task2/Action.kt
package com.chocomiruku.homework1.task2

sealed class Action {
    object Registration : Action()
    class Login(val user: User) : Action()
    object Logout : Action()
}

fun doAction(action: Action) {
    when (action) {
        is Action.Registration -> println("\nRegistration has started.")
        is Action.Login -> {
            println("Auth has started.")
            auth(action.user, ::updateCache)
        }
        is Action.Logout -> print("Logout has started.")
    }
}

inline fun auth(user: User, func: () -> Unit) =
    try {
        user.checkAge()
        anonymousAuthCallback.authSuccess()
        func()
    } catch (e: Exception) {
        println(e.message)
        anonymousAuthCallback.authFailed()
    }

fun updateCache() {
    println("Cache was updated.")
}